### ISSUE TITLE
Chapter 7: There are a few notable exceptions: splitter is absent (forced to "random")

### ISSUE THREAD

**Issue by tyeoyj** [ageron/handson-ml2#340]
Hi Aurelien,

Just want to start off by saying that your book is great and has helped my tremendously in my ML journey.

Just one question: 
In chapter 7 foot note 11, you mention that for a **Random Forest classifier** sets the **Decision Tree classifier** hyperparameter 'splitter' to 'random'.

However, what I understand from reading the subsequent section on **Extra-Trees**, the **Random Forest classifier** decision trees randomly sample a subset of the features but still select the best split among all the features sampled while the **Extra-Trees** are the ones that randomly sample the features followed by a random split.

If that's the case then the bagging classifier that is equivalent of the random forest classifier should be as follows:

```
BaggingClassifier( DecisionTreeClassifier(splitter="best", max_leaf_nodes=16), n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, max_features='sqrt')

```
with splitter='best' and max_features='sqrt' being the changes.

Do let me know what you think, thank you!

**Comment by ageron** [comment]
Hi @tyeoyj ,

Thanks for your kind message, I'm really glad my book was helpful to you. 😊

It took me a while to double-check this, but you're absolutely right.

The documentation for the `splitter` hyperparameter of the `DecisionTreeClassifier` class says:

> splitter{“best”, “random”}, default=”best”
> The strategy used to choose the split at each node. Supported strategies are “best” to choose the best split and “random” to choose the best random split.

I misunderstood the `"random"` strategy as meaning "randomly selecting a subset of features", but after looking at the [source code](https://github.com/scikit-learn/scikit-learn/blob/95119c13af77c76e150b753485c662b7c52a41a2/sklearn/tree/_splitter.pyx#L572), it means that the algorithm will randomly select a threshold for each feature, which is what the `ExtraTrees` class does.

The `"best"` strategy does indeed select a subset of features randomly (which is what I thought `splitter="random"` did).

Moreover, the `max_features` hyperparameter of the `RandomForestClassifier` class is set to `"auto"` by default, which is the same as `"sqrt"`, which computes the square root of the number of features. Sadly, the `max_features` hyperparameter of the `BaggingClassifier` class does not support `"auto"` or `"sqrt"`, only an int or a float, so we have to compute `max_features` manually.

I also noticed that the `presort` hyperparameter of the `BaggingClassifier` class was deprecated in Scikit-Learn 0.22, and it was removed in 0.24 (i.e., there's no possibility of presorting anymore). And the `max_samples` hyperparameter was added to `RandomForestClassifier` in 0.22. A couple more things to fix in this cursed footnote! 😬 

To summarize, the footnote 11 is wrong, I've just removed it. Plus, I've changed the code at the top of page 198 to be:

```python
from math import sqrt

bag_clf = BaggingClassifier(
    DecisionTreeClassifier(max_leaf_nodes=16),
    n_estimators=500, n_jobs=-1, max_features=round(sqrt(X_train.shape[1])))
```

instead of:

```python
bag_clf = BaggingClassifier(
    DecisionTreeClassifier(splitter="random", max_leaf_nodes=16),
    n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1)
```

I added a new footnote in the Extra-Trees paragraph: "You can create a single extremely random tree by setting `splitter="random"` when creating a `DecisionTreeClassifier`."

Thanks so much for catching this error! 👍
